accepted/exception-handling.md

```diff
@@ -0,0 +1,240 @@
+
+* The official W3C WebAssembly exception handling revised proposal;
+* and, the legacy exception handling revision.
+
+The revised proposal was advanced to phase 4 during [a community group
+meeting in
+July](https://github.com/WebAssembly/meetings/blob/main/main/2024/CG-07-16.md). Support
+for exception handling is (at least) of interest to C++, Kotlin, and
+OCaml toolchains. The proposal is also a prerequisite for the [stack
+switching
+proposal](https://github.com/WebAssembly/stack-switching/blob/main/proposals/stack-switching/Explainer.md)
+which plans to use exceptions to finalise stacks. Nonetheless, to
+remain standard compliant, we will eventually have to implement the
+exception handling proposal in Wasmtime.
+
+# Requirements
+[requirements]: #requirements
+
+* In its end state exception handling should be a zero-cost feature,
+  i.e. the implementation must not affect the run-time performance
+  (execution speed, memory usage, etc) of programs that do not use the
+  feature. However, in the interim we are willing to relax this
+  requirement to a "near zero-cost" such that we can use a simple
+  implementation to quickly enable users to use this feature.
+
+* No dependence on `libunwind`. This library provides a C API for
+  determining and unwinding call chains in ELF programs. Nonetheless,
+  we do not want to bring it into our trusted computing base (TCB),
+  because we have found its implementations to be buggy in the past,
+  and such bugs can easily compromise the integrity of Wasmtime (some
+  past `libunwind` bugs:
+  [#2808](https://github.com/bytecodealliance/wasmtime/issues/2808),
+  [#3256](https://github.com/bytecodealliance/wasmtime/issues/3256),
+  [#7997](https://github.com/bytecodealliance/wasmtime/issues/7997)).
+
+* Fast unwinding strategy. We require unwinding to be fast enough to
+  support [the guest
+  profiler](https://docs.wasmtime.dev/examples-profiling-guest.html). Thus,
+  it is important that the chosen implementation strategy does not box
+  ourselves into a corner. Though, we consider it acceptable if the
+  minimal viable product (MVP) of the implementation does not achieve
+  peak performance for unwinding as long as there is a viable path to
+  improving the performance.
+
+* Compatibility with standard formats. Regardless of unwinding
+  strategy, Wasmtime must continue to work with system profilers which
+  suport DWARF or frame pointers, e.g. `perf`. Furthermore, in
+  Cranelift `cg_clif` should be extended with capabilities to leverage
+  the unwind info support to emit appropriate DWARF, Windows
+  Structured Exception Handling, and so forth. Though, we believe a
+  MVP capable of recovering the `vmctx` register should suffice for
+  critical use-cases.
+
+# Non-requirements
+[non-requirements]: #non-requirements
+
+* No support for the legacy exception handling
+  revision. Justification: the legacy revision is being phased out.
+
+* No support for unwinding across host frames. Justification:
```

Comment:
One could add to the justification that it is, in the general case, _impossible_ to unwind arbitary host code. Examples:
1) .NET Jitted host code - it won't have any unwind info Wasmtime could hope to understand (.NET runtime uses an internal Windows-like unwind info even on Unix OSes).
2) C code compiled without unwind info (and with omitted frame pointers, for good measure).
3) Windows x86 native code does not support virtual unwinding at all.